### PR TITLE
[Android P] Cutout support

### PR DESCRIPTION
**Pull request by lgvalle**
This PR adds an Android P Spikes project to host spikes for small features and the first of those spikes: an activity demonstrating how the new Cutout support works

There's a `README.md` included explaining how this cutout support works. I'm duplicating the content of that file here for the benefit of people reviewing this PR:

## [Display coutout support](https://developer.android.com/preview/features.html#cutout)

Notches are called Display Cutouts on Android and Android P is adding support for them

In Android P you can simulate a display coutout with three different form factors.

> Tall display cutout

When your activity screen goes behind the cutout you can specify how it should behave:

* [`LAYOUT_IN_DISPLAY_CUTOUT_MODE_DEFAULT` ](https://developer.android.com/reference/android/view/WindowManager.LayoutParams.html#LAYOUT_IN_DISPLAY_CUTOUT_MODE_DEFAULT)

> The window is allowed to extend into the DisplayCutout area, only if the DisplayCutout is fully contained within the status bar. Otherwise, the window is laid out such that it does not overlap with the DisplayCutout area.

> In practice, this means that if the window did not set FLAG_FULLSCREEN or SYSTEM_UI_FLAG_FULLSCREEN, it can extend into the cutout area in portrait. Otherwise (i.e. fullscreen or landscape) it is laid out such that it does overlap the cutout area.
> 
> The usual precautions for not overlapping with the status bar are sufficient for ensuring that no important content overlaps with the DisplayCutout.

* [`LAYOUT_IN_DISPLAY_CUTOUT_MODE_ALWAYS`](https://developer.android.com/reference/android/view/WindowManager.LayoutParams.html#LAYOUT_IN_DISPLAY_CUTOUT_MODE_ALWAYS)

> The window is always allowed to extend into the DisplayCutout area, even if fullscreen or in landscape.

> The window must make sure that no important content overlaps with the DisplayCutout.

* [`LAYOUT_IN_DISPLAY_CUTOUT_MODE_NEVER`](https://developer.android.com/reference/android/view/WindowManager.LayoutParams.html#LAYOUT_IN_DISPLAY_CUTOUT_MODE_NEVER)

> The window is never allowed to overlap with the DisplayCutout area.
> This should be used with windows that transiently set SYSTEM_UI_FLAG_FULLSCREEN to avoid a relayout of the window when the flag is set or cleared.

The new [DisplayCutout](https://developer.android.com/reference/android/view/DisplayCutout.html) class lets you find out the location and shape of the non-functional areas where content shouldn't be displayed. 

For example, if a view in your layout falls behind a cutout area you can get the bounds of it and layout your view below the cutout

```kotlin
titleView.setOnApplyWindowInsetsListener { view, windowInsets ->
	windowInsets.displayCutout?.let {
	    val safeInsetTop = windowInsets.displayCutout.safeInsetTop
	    titleView.y = titleView.top + safeInsetTop.toFloat()	    
	}
	windowInsets
}
```

## Screenshots

 Cutout support | 
 --- | 
cutout_support |

### CODE REVIEW

**Comment by lgvalle** [comment]
The PR looks large because there's a bunch of noisy files. The actual code changes are small

File: AndroidP/.gitignore [hunk 1]
```diff
@@ -0,0 +1,9 @@
+*.iml
+.gradle
+/local.properties
+/.idea/workspace.xml
+/.idea/libraries
+.DS_Store
+/build
+/captures
+.externalNativeBuild
```
Comment (rock3r) [inline AndroidP/.gitignore]: You sure this is needed? You may want to look at [this](https://github.com/novoda/spikes/blob/54b73a7ebe60aa296c9cbe76d64b49783781340d/sliceanddice/.gitignore) `.gitignore` instead

File: AndroidP/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+
+> Tall display cutout
+
+
+When your activity screen goes behind the cutout you can specify how it should behave:
+
+* [`LAYOUT_IN_DISPLAY_CUTOUT_MODE_DEFAULT` ](https://developer.android.com/reference/android/view/WindowManager.LayoutParams.html#LAYOUT_IN_DISPLAY_CUTOUT_MODE_DEFAULT)
+
+> The window is allowed to extend into the DisplayCutout area, only if the DisplayCutout is fully contained within the status bar. Otherwise, the window is laid out such that it does not overlap with the DisplayCutout area.
```
Comment (rock3r) [inline AndroidP/README.md]: Could you indent all these blockquotes by 4 spaces, so they show _inside_ the bullet items instead of outside? Will greatly help readability

File: AndroidP/app/.gitignore [hunk 3]
```diff
@@ -0,0 +1 @@
+/build
```
Comment (rock3r) [inline AndroidP/app/.gitignore]: YAGNI if you change line 7 (`/build`) in the root gitignore to `build/` — notice the lack of path separator at the beginning, meaning it applies in sub-folders too

File: AndroidP/app/build.gradle [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+apply plugin: 'kotlin-android'
+
+apply plugin: 'kotlin-android-extensions'
+
+android {
+    compileSdkVersion 'android-P'
+    defaultConfig {
+        applicationId "com.novoda.androidp"
+        minSdkVersion 27
+        targetSdkVersion 28
```
Comment (rock3r) [inline AndroidP/app/build.gradle]: Yay the hack 💃 They should really document this :P Or maybe they don't want us to know...

File: AndroidP/app/build.gradle [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        versionCode 1
+        versionName "1.0"
+        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
```
Comment (rock3r) [inline AndroidP/app/build.gradle]: I'd take out all the test classes and references since we don't have tests, it's just more noise

File: AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt [hunk 6]
```diff
@@ -0,0 +1,97 @@
+package com.novoda.androidp.pt651_display_cutout
```
Comment (rock3r) [inline AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt]: Why this weird package name? What does it mean?

File: AndroidP/app/src/main/java/com/novoda/androidp/MainActivity.kt [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+import android.view.View
+import com.novoda.androidp.pt651_display_cutout.CutoutActivity
+
+class MainActivity : AppCompatActivity() {
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+
+        findViewById<View>(R.id.btn_cutout_activity).setOnClickListener {
```
Comment (rock3r) [inline AndroidP/app/src/main/java/com/novoda/androidp/MainActivity.kt]: You could use the Kotlin Android extensions:

```kotlin
btn_cutout_activity.setOnClickListener { ... }
```

File: AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_cutout)
+        linearLayout = findViewById(R.id.main_layout)
+
+        val pref = PreferenceManager.getDefaultSharedPreferences(this)
+        setLayoutCutoutMode(pref)
+        setupTitleView(pref)
+        setupRadioButtonForCutoutMode(pref)
+        setupRadioButtonForInset(pref)
+
```
Comment (rock3r) [inline AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt]: Nit: extra whiteline

File: AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    private fun setLayoutCutoutMode(pref: SharedPreferences) {
+        window.attributes.layoutInDisplayCutoutMode = cutoutModeMap[pref.getInt(CUTOUT_MODE, CUTOUT_MODE_DEFAULT)] ?: CUTOUT_MODE_DEFAULT
```
Comment (rock3r) [inline AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt]: this is a looong line :D split up?

File: AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+                    val safeInsetTop = windowInsets.displayCutout.safeInsetTop
+                    titleView.y = titleView.top + safeInsetTop.toFloat()
+                }
+            }
+
+            windowInsets
+        }
+    }
+
+    private fun isUseInsetSelected(pref: SharedPreferences) = cutoutUseInsetMap[pref.getInt(CUTOUT_USE_INSET, 0)] ?: false
```
Comment (rock3r) [inline AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt]: Ditto, long line, split up please? Or, newline after the `=`

File: AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+    val cutoutUseInsetMap = mapOf(
+            R.id.rd_cutout_yes to true,
+            R.id.rd_cutout_no to false
+    )
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_cutout)
+        linearLayout = findViewById(R.id.main_layout)
```
Comment (rock3r) [inline AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt]: You can directly use `main_layout` that the Kotlin Android extensions give ya

File: AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        }
+    }
+
+    private fun isUseInsetSelected(pref: SharedPreferences) = cutoutUseInsetMap[pref.getInt(CUTOUT_USE_INSET, 0)] ?: false
+
+    private fun setupRadioButtonForCutoutMode(pref: SharedPreferences) {
+        val radioGroup = findViewById<RadioGroup>(R.id.rd_cutout_mode)
+        radioGroup.check(pref.getInt(CUTOUT_MODE, R.id.rd_cutout_default))
+        radioGroup.setOnCheckedChangeListener { radioGroup, checkedId ->
+            pref.edit().putInt(CUTOUT_MODE, checkedId).commit()
```
Comment (rock3r) [inline AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt]: Why not simply adding an intent extra when restarting and reading it in `onCreate`? Seems less of a roundtrip

File: AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        val radioGroup = findViewById<RadioGroup>(R.id.rd_cutout)
+        radioGroup.check(pref.getInt(CUTOUT_USE_INSET, R.id.rd_cutout_no))
+        radioGroup.setOnCheckedChangeListener { radioGroup, checkedId ->
+            pref.edit().putInt(CUTOUT_USE_INSET, checkedId).commit()
+            restart()
+        }
+    }
+
+    private fun restart() {
+        linearLayout.requestLayout();
```
Comment (rock3r) [inline AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt]: Why is this needed?

File: AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+        startActivity(getIntent());
+        finish();
+    }
+
+    override fun onWindowFocusChanged(hasFocus: Boolean) {
```
Comment (rock3r) [inline AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt]: Not sure why you need this?

File: AndroidP/app/src/main/res/layout/activity_cutout.xml [hunk 15]
```diff
@@ -0,0 +1,73 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (rock3r) [inline AndroidP/app/src/main/res/layout/activity_cutout.xml]: Y U no `ConstraintLayout`?

File: AndroidP/build.gradle [hunk 16]
```diff
@@ -0,0 +1,27 @@
+// Top-level build file where you can add configuration options common to all sub-projects/modules.
+
+buildscript {
+    ext.kotlin_version = '1.2.20'
```
Comment (rock3r) [inline AndroidP/build.gradle]: `1.2.30`?

File: AndroidP/build.gradle [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+    repositories {
+        google()
+        jcenter()
+    }
+    dependencies {
+        classpath 'com.android.tools.build:gradle:3.0.0'
```
Comment (rock3r) [inline AndroidP/build.gradle]: Don't be shy, go `3.1.0` or higher — you theoretically need it to develop for Android P (along with AS 3.1+)

File: AndroidP/gradle/wrapper/gradle-wrapper.properties [hunk 18]
```diff
@@ -0,0 +1,6 @@
+#Fri Mar 09 14:43:16 GMT 2018
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.1-all.zip
```
Comment (rock3r) [inline AndroidP/gradle/wrapper/gradle-wrapper.properties]: 4.6? 💃

File: AndroidP/README.md (hunk 2 above)
Comment (lgvalle) [reply AndroidP/README.md]: Didn't know you can do that 💯

File: AndroidP/app/src/main/res/layout/activity_cutout.xml (hunk 15 above)
Comment (lgvalle) [reply AndroidP/app/src/main/res/layout/activity_cutout.xml]: I want to live a happy life

Comment (rock3r) [reply AndroidP/app/src/main/res/layout/activity_cutout.xml]: Buuuh :D

**Comment by rock3r** [comment]
Thanks for the update, still a couple of questions/comments left unanswered, then good to go for me!

File: AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt (hunk 6 above)
Comment (lgvalle) [reply AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt]: `pt651` is the task id in our Jira for this. I seemed like a good idea at that point :/

Comment (rock3r) [reply AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt]: I couldn't have ever understood it, like ever XD Can we switch to something more meaningful? :D external people don't even have access to our JIRA, and this is public

File: AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt (hunk 12 above)
Comment (lgvalle) [reply AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt]: I want to persist it so I can completely kill the app and start it in the selected mode I want. This way I know it's working fine from scratch

Comment (rock3r) [reply AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt]: You could simply use [`persistableMode`](https://developer.android.com/reference/android/R.attr.html#persistableMode) then

File: AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt (hunk 6 above)
Comment (lgvalle) [reply AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt]: Yep I'll drop the `pt651` entirely 👍

File: AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt (hunk 12 above)
Comment (lgvalle) [reply AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt]: Oh wow that's powerful, I didn't know about it. Let me try it 💯

Comment (lgvalle) [reply AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt]: Ok it's not that straight forward because I'm re-launching the same activity.
The problem I'm facing is `onSaveInstanceState` on the first instance of the activity is only called _after_ the second one is launched, so I don't have a `persistentState` to read stuff from.

If I make the activity `singleTask` so there's only one instance then `onSaveInstanceState` is never called because the instance is never destroyed.

Overall I see too much hassle for something that's not the important thing to showcase here :/

Comment (rock3r) [reply AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt]: Yeah becoming way too much hassle — to me it's a big YAGNI to save the state in the preferences in the first place anyway
